CMake Changes

Saito720 opened the pull request:
## Description

This commit revises numerous CMake configurations and the Windows cross-compilation toolchain. These revisions primarily enhance consistency and maintainability across the build process.

### Key Functional Changes

- **Executable Naming**: To maintain consistency with other executables and prevent overwriting the original game files, the `Launcher` and `AmnesiaDemo` executables are now appended with "CE."
- **Static Linking**: All executables compiled with MinGW are statically linked with the GCC and C++ runtime, enhancing portability and reducing dependencies on target system libraries.

### Additional Updates

- Improved the formatting and clarity of comments within the CMake files.
- Reorganized certain sections of the CMake files to enhance clarity and ease future maintenance.

Saito720 committed:
CMake Changes

jkulawik commented:
Shouldn't one `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -static-libgcc -static-libstdc++")` (in core) be enough? It's repeated 3 times

jkulawik commented on HPL2/core/CMakeLists.txt:
```diff
@@ -4,24 +4,27 @@ project(HPL2)
-add_definitions(-DUSE_SDL2)
-add_definitions(-DUSE_OALWRAPPER)
+# We use SDL2 by default in CMake now, but SDL 1.2 is also supported in the code. We need to instruct it to use SDL2.
+add_definitions(-DUSE_SDL2 -DUSE_OALWRAPPER)
```
Might as well change this to `add_compile_definitions(USE_SDL2 USE_OALWRAPPER)` because `add_definitions` is kinda deprecated

Saito720 replied:
Done.

jkulawik commented on HPL2/core/CMakeLists.txt:
```diff
@@ -175,21 +175,20 @@ if(LINUX)
     FindPrebuiltLibrary(GLEW_LIBRARY glew32)
     FindPrebuiltLibrary(DEVIL_LIBRARY DevIL)
 
+    # Add AngelScript subdirectory and link with HPL2
```
Redundant comment but it's ok

Saito720 replied:
Removed.

jkulawik commented on HPL2/tools/CMakeLists.txt:
```diff
@@ -7,9 +7,7 @@ function(AddToolTarget target)
-    add_definitions(
-        -DUSERDIR_RESOURCES
-    )
+    add_definitions(-DUSERDIR_RESOURCES)
```
As previously with `add_definitions`

Saito720 replied:
Done.

jkulawik commented on amnesia/src/CMakeLists.txt:
```diff
@@ -8,71 +8,71 @@ option(USERDIR_RESOURCES "Allow using local userdirectory resources" ON)
-# MinSizeRel: same as Release but optimizing for size rather than speed.
+# Release: High optimization level, no debug info, no debug code, no asserts.
+# Debug: No optimization, asserts enabled, debug info included (enables stepping through code with a debugger and source-file:line-number translation).
+# RelWithDebInfo: Optimized with debug info, but no debug output code or asserts.
```
This makes it sound like debug info is something that optimizes programs

jkulawik replied:
i.e. "optimized with debug info"

Saito720 replied:
Done.

jkulawik commented on amnesia/src/CMakeLists.txt:
```diff
@@ -8,71 +8,71 @@ option(USERDIR_RESOURCES "Allow using local userdirectory resources" ON)
 
 if(MINGW)
-    # There is some stuff that breaks in mingw in newer standards
+    # There are some issues that break in MinGW with newer standards.
```
If we're being pedantic about grammar, it's not issues which break

Saito720 replied:
Done.

jkulawik commented on amnesia/src/CMakeLists.txt:
```diff
@@ -8,71 +8,71 @@ option(USERDIR_RESOURCES "Allow using local userdirectory resources" ON)
-    add_definitions(
-        -DUSERDIR_RESOURCES
-    )
+    add_definitions(-DUSERDIR_RESOURCES)
```
Ditto

Saito720 replied:
Done.

jkulawik commented on amnesia/src/XCompileToWindows.toolchain:
```diff
@@ -1,26 +1,29 @@
-set(CMAKE_CXX_COMPILER i686-w64-mingw32-g++-win32)
+# Which compilers to use for C and C++
+# Uncomment these lines for 64-bit compilation, though these are set for a 32-bit Windows environment
+# set(CMAKE_C_COMPILER   x86_64-w64-mingw32-gcc-win32)  # 64-bit GCC compiler, targeting 32-bit Windows
```
Redundant to say that this is GCC/G++ (it's in the compiler name). Also I don't think it's targeting 32 bit if it's explicitly a 64 bit compiler?

Saito720 replied:
Removed.

jkulawik commented on amnesia/src/XCompileToWindows.toolchain:
```diff
@@ -1,26 +1,29 @@
+set(CMAKE_C_COMPILER   i686-w64-mingw32-gcc-win32)     # 32-bit GCC compiler
+set(CMAKE_CXX_COMPILER i686-w64-mingw32-g++-win32)     # 32-bit G++ compiler
+
+# These options force 32-bit compilation, which shouldn't be necessary as the compilers specified are already for 32-bit
```
Again, is this true? As far as I'm aware `x86_64-w64-mingw32-*-win32` are for 64 bit and `win32` is just the platform name for windows in general. There is no `mingw64` or `-win64` compilers in mingw.

Saito720 replied:
Likely due to confusion on my end.

Removed.

jkulawik reviewed (commented):
Here's a review with several comments inside

Saito720 commented:
> Shouldn't one `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -static-libgcc -static-libstdc++")` (in core) be enough? It's repeated 3 times

No, the three instances are required (as explained on Discord) due to there being three separate CMake configurations that build all the binaries.

Saito720 committed:
Update CMakeLists.txt

Addressed all comments for the HPL2 core CMake.

Saito720 committed:
Update CMakeLists.txt

Addressed comment.

Saito720 committed:
Update CMakeLists.txt

Addressed comments.

Saito720 committed:
Update XCompileToWindows.toolchain

Addressed comments.

jkulawik commented:
Did you test this? `add_compile_definitions(-DUSERDIR_RESOURCES)`
I've had issues with keeping the `-D` prefix in this function before and had to write it like this `add_compile_definitions(USERDIR_RESOURCES)`

Saito720 committed:
Merge branch 'TiManGames:main' into main

Saito720 committed:
Update CMakeLists.txt

Saito720 committed:
Update CMakeLists.txt

Saito720 committed:
Update CMakeLists.txt

Saito720 committed:
Update CMakeLists.txt

Forgot it remove the prefixed "D".

Also yeah, GitHub (website) doesn't allow me to revert the previous commit so I can combine something as insignificant as this into a single commit.

Saito720 commented:
> Did you test this? `add_compile_definitions(-DUSERDIR_RESOURCES)` I've had issues with keeping the `-D` prefix in this function before and had to write it like this `add_compile_definitions(USERDIR_RESOURCES)`

The appended "-D" did indeed cause compilation issues. So those were addressed and now both Windows and Linux compiles as expected.

I also resolved the one remaining warning for Linux compilation.

Windows:

![VirtualBoxVM_UGs7hFnpiH](https://github.com/TiManGames/AmnesiaTheDarkDescent/assets/82419801/7cf183d6-35f9-4544-aa27-c50a142ad4b5)

Linux:
![VirtualBoxVM_JxWeFNDR60](https://github.com/TiManGames/AmnesiaTheDarkDescent/assets/82419801/42b38d95-00f9-4c15-a175-29b704eef497)

jkulawik commented on amnesia/src/CMakeLists.txt:
```diff
@@ -48,6 +48,19 @@ add_compile_options(
 
+# Define LINUX if we are on a Linux system
+if(UNIX AND NOT APPLE)
+    set(LINUX TRUE)
```
This variable is normally set in BoilerPlate and I'm not sure duplicating it here is a good idea - I assume it was added for the warning below and not because it does not propagate to the launcher or the game?

You can just add the warning suppression for all builds along with the previous ones. It's a single occurrence of that warning too, so should be easy to fix in the source code later on.

jkulawik replied:
Sidenote, I do feel like BoilerPlate stuff should be properly available in a global context. At some point I will review it and make it global and set in the src build instead of the hpl2 build, but it's another pet peeve edit that could wait until more important things are resolved - such as the MSVC build or building all libs from source on MinGW...

Saito720 committed:
Update CMakeLists.txt

Removed Linux conditional for warning suppression.

jkulawik commented:
For the future, please don't mix broad cosmetic changes/refactors with functional changes.
